README.md
Language Identifier
===================

![screenshot](media/screenshot.png)

Identification of chosen languages from a live audio stream
 
This code base has been developed by [ZKM | Hertz-Lab](https://zkm.de/en/about-the-zkm/organization/hertz-lab) as part of the project [»The Intelligent Museum«](#the-intelligent-museum). 

Please raise issues, ask questions, throw in ideas or submit code, as this repository is intended to be an open platform to collaboratively improve language identification.

Copyright (c) 2021 ZKM | Karlsruhe.  
Copyright (c) 2021 Paul Bethge.  
Copyright (c) 2021 Dan Wilcox.  

BSD Simplified License.

Dependencies
------------

* [openFrameworks](https://openframeworks.cc/download/)
* openFrameworks addons:
  - ofxOSC (included with oF)
  - [ofxTensorFlow2](https://github.com/zkmkarlsruhe/ofxTensorFlow2)
* [CLI11 parser](https://github.com/CLIUtils/CLI11): included in `src`
* [Language Identification](https://github.com/zkmkarlsruhe/language-identification): trained neural networks placed in `bin/data` 

Tested Platforms
----------------

* MacBook Pro 2017, macOS 10.15 & openFrameworks 0.11.2
* MacBook Pro 2018, macOS 11.3.1 & openFrameworks 0.11.2

Structure
---------

* src/: contains the openFrameworks C++ code
* bin/data/model_*: contains the SavedModels trained with TensorFlow2

Installation & Build
--------------------

Overview:

1. Follow the steps in the ofxTensorFlow2 "Installation & Build" section for you platform
2. Generate the project files for this folder using the OF ProjectGenerator
3. Build for your platform

### Generating Project Files

Project files are not included so you will need to generate the project files for your operating system and development environment using the OF ProjectGenerator which is included with the openFrameworks distribution.

To (re)generate project files for an existing project:

1. Click the "Import" button in the ProjectGenerator
2. Navigate to the project's parent folder ie. "apps/myApps", select the base folder for the example project ie. "LanguageIdentifier", and click the Open button
3. Click the "Update" button

If everything went Ok, you should now be able to open the generated project and build/run the example.

### macOS

On macOS, a couple of additional manual steps are required to use ofxTensorflow2:

1. Enable C++14 in openFrameworks (only once, Xcode + Makefile)
2. Invoke `macos_install_libs.sh` in the Xcode project's Run Script build phases (after every project regeneration, Xcode only)

See the detailed steps in the [ofxTensorflow2 readme](https://github.com/zkmkarlsruhe/ofxTensorFlow2#macos).

For an Xcode build, open the Xcode project, select the "LanguageIdentifier Debug" scheme, and hit "Run".

For a Makefile build, build and run on the terminal:

```shell
cd LanguageIdentifier
make ReleaseTF2
make RunRelease
```
### Linux

Build and run on the terminal:

```shell
cd LanguageIdentifier
make Release
make RunReleaseTF2
```

Usage
-----

The openFrameworks application runs the language identification model using audio input. The detection status and detected language is sent out using OSC (Open Sound Control) messages.

### Key Commands

* `l`: toggle start/stop listening
* `a`: toggle listening auto stop after detection

### OSC Communication

#### Sending

By default, sends to:
* address: `localhost` ie. `127.0.0.1`
* port: `9999`

Message specification:

* **/detected _status_**: detection status
  - status: float, boolean 1 found - 0 lost
* **/lang _index_ _name_ _confidence_**: detected language
  - index: int, language map index
  - name: string, language map name
  - confidence: float, confidence percentage 0 - 100

#### Receiving

By default, listens on:
* port `9898`

Message specification:

* **/listen**: start listening
* **/listen _state_**: start/stop listening
  - state: bool, 0 - stop, 1 - start
* **/autostop**: enable listening auto stop after detection
* **/autostop _state_**: enable/disable listening auto stop after detection
  - state: bool, 0 - keep listening, 1 - stop on detection

### Commandline Options

Additional run time settings are available via commandline options as shown via the `--help` flag output:

```shell
% bin/LanguageIdentifier --help
identifies spoken language from audio stream
Usage: LanguageIdentifier [OPTIONS]

Options:
  -h,--help                   Print this help message and exit
  -s,--senders TEXT ...       OSC sender addr:port host pairs, ex. "192.168.0.100:5555" or multicast "239.200.200.200:6666", default "localhost:9999"
  -p,--port INT               OSC receiver port, default 9898
  -c,--confidence FLOAT:FLOAT bounded to [0 - 1]
                              min confidence, default 0.75
  -t,--threshold FLOAT:INT bounded to [0 - 100]
                              volume threshold, default 25
  -l,--list                   list audio input devices and exit
  --inputdev INT              audio input device number
  --inputname TEXT            audio input device name, can do partial match, ex. "Microphone"
  --inputchan INT             audio input device channel, default 1
  -r,--samplerate INT         audio input device samplerate, can be 441000 or a multiple of 16000, default 48000
  --nolisten                  do not listen on start
  --autostop                  stop listening automatically after detection
  -e,--execute TEXT           command to execute on detection with key=value pair args
  -v,--verbose                verbose printing
  --version                   print version and exit
```

For example, to send OSC to multiple addresses use the `-s` option:

```shell
% bin/LanguageIdentifier -s localhost:9999 localhost:6666 192.168.0.101:7777
```

#### macOS

For macOS, the application binary can be invoked from within the .app bundle to pass commandline arguments:

```shell
bin/LanguageIdentifier.app/Contents/MacOS/LanguageIdentifier -h
```

This approach can also be wrapped up into a shell alias to be added to the account's `~/.bash_profile` or `~/.zshrc` file:

```
alias langid="/Applications/LanguageIdentifier.app/Contents/MacOS/LanguageIdentifier"
```

Reload the shell and application can now be invoked via:

```shell
% langid -v --inputdev 2
```

### Executing commands

LanguageIdentifier can execute a command on a language detection via the `-e/--execute` option:

```shell
% bin/LanguageIdentifier -e `pwd`/script.sh
```

The arguments passed to the command are a series of key & values pairs where `key=value`:

* selected key: string value, name of detected language
* LANG key: float value, normalized detection confidence 0-1 for each supported language

Example args:

~~~
selected=noise noise=0.753628 chinese=0.000086 english=0.237782 french=0.001154 german=0.002538 italian=0.000791 russian=0.000018 spanish=0.004004
~~~  

_Note: In general, the command must include the full path if it is not in current shell PATH._

Demos
-----

The demos consist of rapid prototypes built using the following components:

* language identifier
* visual front end: [loaf](http://danomatika.com/code/loaf)

Custom visual front ends are written in Lua for [loaf](http://danomatika.com/code/loaf), a Lua interpreter with bindings for [openFrameworks](http://openframeworks.cc/) which includes a built-in Open Sound Control (OSC) server.

### Usage

To set up a run environment on macOS, download loaf and place the .app in the system `/Applications` folder.

To run a loaf project, drag the main Lua script or project folder onto the loaf.app.

Notes
-----

### Sample Rate

The model inputs audio with a sample rate of 16 kHz, so the incoming stream is downsampled and the app's input sample rate needs to be a multiple of 16, ie. 48kHz, 92kHz, etc.

As 44.1kHz is also common, it is accepted and treated as 48kHz but the downsampled audio is then higher in pitch and may be noisy. In our tests, however detection is still acceptable.

Develop
-------

### Release steps

1. Update changelog
2. Update app version in Xcode project and ofApp.h define
3. Tag version commit, ala "0.3.0"
4. Push commit and tags to server:

    git commit push
    git commit push --tags

The Intelligent Museum
----------------------

An artistic-curatorial field of experimentation for deep learning and visitor participation

The [ZKM | Center for Art and Media](https://zkm.de/en) and the [Deutsches Museum Nuremberg](https://www.deutsches-museum.de/en/nuernberg/information/) cooperate with the goal of implementing an AI-supported exhibition. Together with researchers and international artists, new AI-based works of art will be realized during the next four years (2020-2023).  They will be embedded in the AI-supported exhibition in both houses. The Project „The Intelligent Museum” is funded by the Digital Culture Programme of the [Kulturstiftung des Bundes](https://www.kulturstiftung-des-bundes.de/en) (German Federal Cultural Foundation) and funded by the [Beauftragte der Bundesregierung für Kultur und Medien](https://www.bundesregierung.de/breg-de/bundesregierung/staatsministerin-fuer-kultur-und-medien) (Federal Government Commissioner for Culture and the Media).

As part of the project, digital curating will be critically examined using various approaches of digital art. Experimenting with new digital aesthetics and forms of expression enables new museum experiences and thus new ways of museum communication and visitor participation. The museum is transformed to a place of experience and critical exchange.

![Logo](media/Logo_ZKM_DMN_KSB.png)


CHANGES.txt
0.3.0: 2022 Feb 21

* fix crash due to wrong prev buffer value

* added audio support for multi-channel device input and settable samplerate
* added OSC receiver on default port 9898
* added manual listening control via OSC /listen message and l key press
* added listening auto stop on detection via OSC /autostop message or a key press
* added optional command execution on detection
* added commandline argument parsing
* added verbose printing

* now use 7 lang model
* don't enable recording on start
* dont draw extra graph start & end vertices
* graph now goes to 0 when listening is stopped

0.2.0: 2021 Jun 23

* added OSC sending
* added on screen recording status indicator
* added newer attrnn default model
* added demo1
* added changelog

* renamed project to "Language Identifier"
* source file format improvements and reorganization

* ignore generated xcode project files

0.1.0: 2021 May 20

initial version for testing


src/Commandline.cpp
/*
 * Language Identifier
 *
 * Copyright (c) 2021 ZKM | Hertz-Lab
 * Paul Bethge <bethge@zkm.de>
 * Dan Wilcox <dan.wilcox@zkm.de>
 *
 * BSD Simplified License.
 * For information on usage and redistribution, and for a DISCLAIMER OF ALL
 * WARRANTIES, see the file, "LICENSE.txt," in this distribution.
 *
 * This code has been developed at ZKM | Hertz-Lab as part of „The Intelligent
 * Museum“ generously funded by the German Federal Cultural Foundation.
 */

#include "Commandline.h"

Commandline::Commandline(ofApp *app) : app(app) {
	parser.description(DESCRIPTION);
}

bool Commandline::parse(int argc, char **argv) {

	// local options, the rest are ofApp instance variables
	std::vector<std::string> senders;
	int port = 0;
	bool list = false;
	int inputNum = -1;
	std::string inputName = "";
	int inputChannel = 0;
	int sampleRate = 0;
	bool nolisten = false;
	bool autostop = false;
	bool verbose = false;
	bool version = false;
	std::string command = "";

	parser.add_option("-s,--senders", senders,
		"OSC sender addr:port host pairs, ex. \"192.168.0.100:5555\" "
		"or multicast \"239.200.200.200:6666\", default \"localhost:9999\"")->expected(-1);
	parser.add_option("-p,--port", port, "OSC receiver port, default " + ofToString(app->port));
	parser.add_option("-c,--confidence", app->minConfidence,
		"min confidence, default " + ofToString(app->minConfidence))->transform(CLI::Bound(0.0, 1.0));
	parser.add_option("-t,--threshold", app->volThreshold,
	    "volume threshold, default " + ofToString(app->volThreshold))->transform(CLI::Bound(0, 100));
	parser.add_flag(  "-l,--list", list, "list audio input devices and exit");
	parser.add_option("--inputdev", inputNum, "audio input device number");
	parser.add_option("--inputname", inputName, "audio input device name, can do partial match, ex. \"Microphone\"");
	parser.add_option("--inputchan", inputChannel, "audio input device channel, default 1");
	parser.add_option("-r,--samplerate", sampleRate, "audio input device samplerate, can be 441000 or a multiple of " +
		ofToString(ofApp::modelSampleRate) +  ", default " + ofToString(app->sampleRate));
	parser.add_flag(  "--nolisten", nolisten, "do not listen on start");
	parser.add_flag(  "--autostop", autostop, "stop listening automatically after detection");
	parser.add_option("-e,--execute", command, "command to execute on detection with key=value pair args");
	parser.add_flag(  "-v,--verbose", verbose, "verbose printing");
	parser.add_flag(  "--version", version, "print version and exit");

	try {
		parser.parse(argc, argv);
	}
	catch(const CLI::ParseError &e) {
		error = e;
		return false;
	}

	// verbose printing?
	ofSetLogLevel(PACKAGE, (verbose ? OF_LOG_VERBOSE : OF_LOG_NOTICE));

	// print version
	if(version) {
		std::cout << VERSION << std::endl;
		return false;
	}

	// list audio input devices
	if(list) {
		auto devices = app->soundStream.getDeviceList();
		std::size_t count = 0;
		std::cout << "input devices (# NAME [CHANNELS]):" << std::endl;
		for(std::size_t i = 0; i < devices.size(); i++) {
			auto device = devices[i];
			if(device.inputChannels > 0) {
				std::cout << i << " " << device.name
						  << " [" << device.inputChannels << "]" << std::endl;
				count++;
			}
		}
		if(count == 0) {
			std::cout << "none" << std::endl;
		}
		return false;
	}

	// set audio input from device number
	if(inputNum >= 0) {
		auto devices = app->soundStream.getDeviceList();
		if(inputNum >= devices.size()) {
			ofLogError(PACKAGE) << "invalid audio device number: " << inputNum;
			error = CLI::RuntimeError("invalid audio device number", EXIT_FAILURE);
			return false;
		}
		ofSoundDevice &device = devices[inputNum];
		if(device.inputChannels == 0) {
			ofLogError(PACKAGE) << "audio device " << inputNum << " has no input channels";
			error = CLI::RuntimeError("audio device has no input channels", EXIT_FAILURE);
			return false;
		}
		app->inputDevice = inputNum;
	}

	// set audio input from device name
	if(inputName != "") {
		inputNum = -1;
		auto devices = app->soundStream.getDeviceList();
		for(std::size_t i = 0; i < devices.size(); ++i) {
			auto device = devices[i];
			if(device.name.find(inputName) != std::string::npos && device.inputChannels > 0) {
				inputNum = i;
				break;
			}
		}
		if(inputNum >= 0) {
			app->inputDevice = inputNum;
		}
		else {
			ofLogWarning(PACKAGE) << "audio input name not found: " << inputName;
		}
	}

	// set audio input channel
	if(inputChannel > 0) {
		app->inputChannel = inputChannel-1; // 1-index to 0-index
	}

	// set audio input rate
	if(sampleRate > 0) {
		bool set = true;
		if(sampleRate == 44100) {
			// treat as 48k default, pitch change is minimal enough to not affect detection
			// and we don't handle non-integer downsampling factors
			app->sampleRate = sampleRate;
			app->downsamplingFactor = 3;
			set = false;
		}
		else if(sampleRate % ofApp::modelSampleRate != 0) {
			ofLogWarning(PACKAGE) << "ignoring input sample rate which is not a multiple of "
			                      << ofApp::modelSampleRate << ": " << sampleRate;
		}
		if(set) {
			app->sampleRate = sampleRate;
			app->downsamplingFactor = sampleRate / ofApp::modelSampleRate;
		}
	}

	// parse sender host strings
	// split string by last : to get address & port pair,
	// handle bracketed IPv6 hostnames: [::1]:8081
	for(auto host : senders) {
		std::size_t found = host.find_last_of(":");
		if(found == std::string::npos) {
			ofLogWarning(PACKAGE) << "ignoring sender host without port: " << host;
			continue;
		}
		std::string addr = host.substr(0, found);
		std::string port = host.substr(found+1);
		if(addr.size() == 0 || port.size() == 0) {
			ofLogWarning(PACKAGE) << "ignoring sender host with empty address or port: " << host;
			continue;
		}
		if(addr[0] == '[' && addr[addr.size()-1] == ']') {
			addr = addr.substr(1, addr.size()-2);
		}
		int p = ofToInt(port);
		if(p <= 1024) {
			ofLogWarning(PACKAGE) << "ignoring sender host with invalid port or system port: " << host;
			continue;
		}
		app->hosts.push_back(ofApp::OscHost(addr, p));
	}
	if(app->hosts.empty()) { // default
		app->hosts.push_back(ofApp::OscHost("localhost", 9999));
	}

	// receiver port
	if(port > 0) {
		if(port <= 1024) {
			ofLogWarning(PACKAGE) << "ignoring receiver port in system range (0-1024): " << port;
		}
		else {
			app->port = port;
		}
	}

	// no listen
	if(nolisten) {
		app->stopListening();
	}

	// auto stop
	if(autostop) {
		app->autostop = true;
	}

	// command
	if(command != "") {
		app->command = command;
	}

	return true;
}

int Commandline::exit() {
	return parser.exit(error);
}
